### PR TITLE
Android Cache Test 

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt [hunk 1]
```diff
@@ -595,6 +598,59 @@ class OkHttpTest {
...
+            assertNull(it.networkResponse)
+          }
+
+      Thread.sleep(5000)
+
+      client.newCall(request)
+          .execute()
+          .use {
+            assertEquals(200, it.code)
+            // TODO understand this?
```
Comment (swankjesse) [inline android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt:641]: It means the network was used to check whether the cache was fresh

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt [hunk 2]
```diff
@@ -595,6 +598,59 @@ class OkHttpTest {
...
+    val cacheSize = 1L * 1024 * 1024 // 1MB
+    val cache = Cache(ctxt.cacheDir.resolve("testCache"), cacheSize)
+
+    try {
+      client = client.newBuilder()
+          .cache(cache)
+          .build()
+
+      val request = Request.Builder()
+          .url("https://httpbin.org/cache/3")
```
Comment (swankjesse) [inline android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt:616]: use MockWebServer here?

Comment (yschimke) [reply android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt:616]: I've previously preferred real world servers to confirm that things are working end to end.  But given the tests are relatively stable, it's probably a good point to switch from manually running to running in Github.

Will copy the approach here https://twitter.com/colinwhi/status/1244426550918823937

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt (hunk 1 above)
Comment (yschimke) [reply android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt:641]: I think this always surprises me, that Caches don't stop the network request got valid content. :)

I need to read up more on our heuristic caching.

https://www.mnot.net/blog/2009/02/24/unintended_caching
https://paulcalvano.com/index.php/2018/03/14/http-heuristic-caching-missing-cache-control-and-expires-headers-explained/

Comment (yschimke) [reply android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt:641]: > It means the network was used to check whether the cache was fresh

Actually I'm convinced it is the opposite case, the cache was fetched, found stale and then a network call was made.

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt (hunk 2 above)
Comment (swankjesse) [reply android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: Hell yes. Yay @colinrtwhite !

Comment (colinrtwhite) [reply android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt]: Nice! Glad this was useful for y'all 😄

**Pull request merged**